src/backend/src/main/java/com/linkedpipes/lpa/backend/DiscoveryController.java

```diff
@@ -0,0 +1,114 @@
+import java.io.IOException;
+import java.net.URI;
+
+
+import jdk.incubator.http.HttpClient;
+import jdk.incubator.http.HttpRequest;
+import jdk.incubator.http.HttpResponse;
+import static jdk.incubator.http.HttpRequest.BodyPublisher.fromString;
+
+@RestController
+public class DiscoveryController {
+
+    private static final Logger logger =
+            LoggerFactory.getLogger(DiscoveryController.class);
+
+    private HttpUrlConnector httpUrlConnector = new HttpUrlConnector();
+
+    private static final HttpClient client = HttpClient.newBuilder()
+            .followRedirects(HttpClient.Redirect.NEVER)
+            .build();
+
+    @RequestMapping("/pipelines/discover")
+    public Integer startDiscovery(@RequestBody DataSourceList dataSourceList){
+        Integer testDiscoveryId = 1;
+        return testDiscoveryId;
+    }
+
+    @RequestMapping("/pipelines/discoverFromInput")
+    public ResponseEntity<?> startDiscoveryFromInput(@RequestBody String discoveryConfig){
+        HttpResponse<String> response;
+
+        try {
+            URI requestUri = new URI(Application.config.getProperty("discoveryServiceUrl") + "/discovery/startFromInput");
+            HttpRequest request = HttpRequest.newBuilder(requestUri)
+                    .header("Accept", "application/json")
+                    .header("Content-Type", "text/plain")
+                    .POST(fromString(discoveryConfig))
+                    .build();
+            response = client.send(request, HttpResponse.BodyHandler.asString());
+        } catch (Exception e) {
+            logger.error("Exception: ", e);
+            return new ResponseEntity(e.toString(), HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        if (response.statusCode() != 200) {
+            String errorMsg = response.statusCode() + ": " + response.body();
+            logger.error(errorMsg);
+            return new ResponseEntity(errorMsg, HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        return ResponseEntity.ok(response.body());
+    }
+
+    @RequestMapping("/pipelines/discoverFromInputIri")
+    public ResponseEntity<String> startDiscoveryFromInputIri(@RequestParam( value="discoveryConfigIri") String discoveryConfigIri) throws IOException{
+        if(discoveryConfigIri == null || discoveryConfigIri.isEmpty()) {
+            return new ResponseEntity("Input IRI not provided", HttpStatus.BAD_REQUEST);
+        }
+
+        String response = httpUrlConnector.sendGetRequest(Application.config.getProperty("discoveryServiceUrl") + "/discovery/startFromInputIri",
+                "?iri=" + discoveryConfigIri, "application/json");
+
+        return ResponseEntity.ok(response);
+    }
+
+    @RequestMapping("/discovery/{id}/status")
+    public ResponseEntity<String> getDiscoveryStatus(@PathVariable("id") String discoveryId) throws IOException{
+        String response = httpUrlConnector.sendGetRequest(Application.config.getProperty("discoveryServiceUrl") + "/discovery/" + discoveryId,
```

Comment:
I would create a component called DiscoveryServiceComponent, with private method get(string urlPath)

and public methods that are strongly typed in their interface, e.g. getDiscovery(Guid discoveryId) and are implemented by calling the get method, which contains the logic of creating the http request, etc.

That's just an advise, but you are copying so much code...

Reply:
https://kamilszymanski.github.io/refactoring-stringly-typed-systems/